Point_set_processing_3/include/CGAL/structure_point_set.h

```diff
@@ -0,0 +1,1483 @@
+        }
+  }
+
+  void resample_planes (double epsilon)
+  {
+    double grid_length = epsilon * (std::sqrt(2.) - 1e-3);
+
+    for (std::size_t c = 0; c < m_planes.size (); ++ c)
+      {
+        //plane attributes and 2D projection vectors
+        Plane plane = static_cast<Plane> (*(m_planes[c]));
+        Vector vortho = plane.orthogonal_vector();
+        Vector b1 = plane.base1();
+        Vector b2 = plane.base2();
+			
+        b1 = b1 / std::sqrt (b1 * b1);
+        b2 = b2 / std::sqrt (b2 * b2);
+
+        std::vector<Point_2> points_2d;
+
+        //storage of the 2D points in "pt_2d"
+        for (std::size_t j = 0; j < m_planes[c]->indices_of_assigned_points ().size(); ++ j)
+          {
+            std::size_t ind = m_planes[c]->indices_of_assigned_points ()[j];
+            const Point& pt = m_points[ind];
+            points_2d.push_back (Point_2 (b1.x() * pt.x() + b1.y() * pt.y() + b1.z() * pt.z(),
+                                          b2.x() * pt.x() + b2.y() * pt.y() + b2.z() * pt.z()));
+          }
+
+
+        //creation of a 2D-grid with cell width = grid_length, and image structures
+        CGAL::Bbox_2 box_2d = CGAL::bbox_2 (points_2d.begin(), points_2d.end());
+        std::size_t Nx = (std::size_t)((box_2d.xmax() - box_2d.xmin()) / grid_length) + 1;
+        std::size_t Ny = (std::size_t)((box_2d.ymax() - box_2d.ymin()) / grid_length) + 1;
+          
+        std::vector<std::vector<bool> > Mask (Nx, std::vector<bool> (Ny, false));
+        std::vector<std::vector<bool> > Mask_border (Nx, std::vector<bool> (Ny, false));
+        std::vector<std::vector<std::vector<std::size_t> > >
+          point_map (Nx, std::vector<std::vector<std::size_t> > (Ny, std::vector<std::size_t>()));
+
+        //storage of the points in the 2D-grid "point_map"
+        for (std::size_t i = 0; i < points_2d.size(); ++ i)
+          {
+            std::size_t ind_x = (std::size_t)((points_2d[i].x() - box_2d.xmin()) / grid_length);
+            std::size_t ind_y = (std::size_t)((points_2d[i].y() - box_2d.ymin()) / grid_length);
+            Mask[ind_x][ind_y] = true;
+            point_map[ind_x][ind_y].push_back (m_planes[c]->indices_of_assigned_points ()[i]);
+          }
+
+        //hole filing in Mask in 4-connexity
+        for (std::size_t j = 1; j < Ny - 1; ++ j)
+          for (std::size_t i = 1; i < Nx - 1; ++ i)
+            if( !Mask[i][j]
+                && Mask[i-1][j] && Mask[i][j-1]
+                && Mask[i][j+1] && Mask[i+1][j] )
+              Mask[i][j]=true;
+					
+        //finding mask border in 8-connexity	
+        for (std::size_t j = 1; j < Ny - 1; ++ j)
+          for (std::size_t i = 1; i < Nx - 1; ++ i)
+            if( Mask[i][j] &&
+                ( !Mask[i-1][j-1] || !Mask[i-1][j] ||
+                  !Mask[i-1][j+1] || !Mask[i][j-1] ||
+                  !Mask[i][j+1] || !Mask[i+1][j-1] ||
+                  !Mask[i+1][j]|| !Mask[i+1][j+1] ) )
+              Mask_border[i][j]=true;
+          
```

Comment:
Just a remark in case it applies in you case. `boost` provides `boost::dynamic_bitset<>` that can do operations on a whole bitset. Not sure it can help here (or if it is critical), but maybe you can skip some for-loops.